src/main/java/com/mongodb/spark/sql/connector/connection/MongoConnectionProvider.java

```diff
@@ -82,5 +144,56 @@ public void doWithCollection(
         mongoDatabase ->
             function.apply(mongoDatabase.getCollection(collectionName, BsonDocument.class)));
   }
-  // CHECKSTYLE.ON:JavadocMethod
+
+  /** @return memoize's and returns the {@link MongoClientCache} */
+  private static synchronized MongoClientCache getMongoClientCache() {
```

Comment:
Creates the singleton MongoClientCache if it doesn't already exist.

Uses the same system property as the existing spark connector for configuring the keep alive period

Reply:
A couple of months ago a similar lazy initialization approach in the BSON library was reported to cause performance problems to a client, and the client [proposed](https://github.com/mongodb/mongo-java-driver/pull/759) to use the [DCL](https://twitter.com/shipilev/status/1229730431223975936?lang=en) (the version mentioned in the linked tweet is rather a "double-locked checking" 😄). Do we want to consider using the DCL here to avoid a similar situation again?

Alternatively, we can remove the method `getMongoClientCache()` and utilize the [Java class lazy initialization rules](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.4.1):
```java
private static final class Lazy {
    static MongoClientCache mongoClientCache;

    static {
        int keepAliveMS = 5000;
        try {
            keepAliveMS =
                    Integer.parseInt(System.getProperty(SYSTEM_MONGO_CACHE_KEEP_ALIVE_MS_PROPERTY, "5000"));
        } catch (NumberFormatException e) {
            // ignore and use default
        }
        mongoClientCache = new MongoClientCache(keepAliveMS);
        Runtime.getRuntime().addShutdownHook(new ShutdownHook(mongoClientCache));
    }
}
```
With the above class, we can replace all invocations of `getMongoClientCache()` with `Lazy.mongoClientCache`, and it will be guaranteed that a `MongoClientCache` is created only when the execution actually tries to read from the `mongoClientCache` field.